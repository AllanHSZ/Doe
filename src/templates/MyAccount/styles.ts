import styled, { css } from 'styled-components'

export const Wrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    flex-direction: column;
    margin-top: ${theme.spacings.xsmall};
    gap: ${theme.spacings.xxsmall};
    max-width: 32rem;

    input {
      width: 100%;
    }
  `}
`

export const Title = styled.h3`
  ${({ theme }) => css`
    margin-top: ${theme.spacings.small};
  `}
`

export const Heading = styled.h5`
  ${({ theme }) => css`
    margin-bottom: ${theme.spacings.xxsmall};
  `}
`

export const Main = styled.main`
  ${({ theme }) => css`
    display: flex;
    flex-direction: column;
    padding: ${theme.spacings.small} 0;
    align-items: flex-end;
    gap: ${theme.spacings.small};
  `}
`

export const Section = styled.section`
  ${({ theme }) => css`
    display: flex;
    flex-direction: column;
    gap: ${theme.spacings.xxsmall};
    width: 100%;
  `}
`
