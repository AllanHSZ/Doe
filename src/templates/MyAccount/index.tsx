import { useState } from 'react'
import { Lock, User as UserIcon } from 'react-feather'
import { useSession } from 'next-auth/client'
import { useRouter } from 'next/router'
import { AxiosError } from 'axios'

import api from 'services/api'
import { UserProps } from 'services/user'
import handlerError from 'utils/handle-error'
import { Form, FormError, FormLoading } from 'components/Form'
import { editUserValidate, FieldErrors } from 'utils/validations'
import { useToast } from 'hooks/use-toast'
import Address from 'components/Address'
import Button from 'components/Button'
import Input from 'components/Input'
import Page from 'templates/Page'

import * as S from './styles'

const MyAccount = (user: UserProps) => {
  const [session] = useSession()
  const showToast = useToast()
  const { push } = useRouter()

  const [loading, setLoading] = useState(false)
  const [formError, setFormError] = useState('')
  const [fieldError, setFieldError] = useState<FieldErrors>({})
  const [values, setValues] = useState({
    name: user.name,
    password: '',
    confirm_password: ''
  })

  if (!user) {
    showToast({ type: 'error', message: 'Usuário não encontrado' })
    push('/')
  }

  const handleInput = (field: string, value: string) => {
    setValues((s) => ({ ...s, [field]: value }))
  }

  const handleSubmit = async (event: React.FormEvent) => {
    event.preventDefault()
    setFormError('')

    const errors = editUserValidate(values)

    if (Object.keys(errors).length) {
      setFieldError(errors)
      return
    }

    setFieldError({})
    setLoading(true)

    try {
      const { data } = await api.put<UserProps>('/users', {
        name: values.name,
        password: values.password
      })

      if (session) {
        session.user = {
          ...session?.user,
          name: data.name
        }
      }

      setValues({
        name: data.name,
        password: '',
        confirm_password: ''
      })

      showToast({ type: 'success', message: 'Alterações salvas com sucesso' })
    } catch (error) {
      setFormError(handlerError(error as AxiosError))
    } finally {
      setLoading(false)
    }
  }

  return (
    <Page header footer>
      <S.Wrapper>
        <h3>Minha conta</h3>
        <S.Main>
          <Form onSubmit={handleSubmit}>
            <S.Section>
              <S.Heading>Dados pessoais</S.Heading>
              <Input
                name="name"
                label="Nome"
                type="text"
                placeholder="Ex: João Silva"
                error={fieldError?.name}
                value={values.name}
                onInputChange={(v) => handleInput('name', v)}
                disabled={loading}
                icon={<UserIcon />}
              />
            </S.Section>

            <S.Section>
              <S.Heading>Alterar senha</S.Heading>
              <Input
                name="password"
                label="Senha"
                placeholder="****"
                error={fieldError?.password}
                onInputChange={(v) => handleInput('password', v)}
                disabled={loading}
                type="password"
                icon={<Lock />}
              />

              <Input
                name="confirm_password"
                label="Confirmar senha"
                error={fieldError?.confirm_password}
                onInputChange={(v) => handleInput('confirm_password', v)}
                placeholder="****"
                type="password"
                disabled={loading}
                icon={<Lock />}
              />
            </S.Section>

            {!!formError && <FormError>{formError}</FormError>}

            <Button
              type="submit"
              disabled={
                loading || (user.name === values.name && !values.password)
              }
            >
              {loading ? <FormLoading /> : <span>SALVAR</span>}
            </Button>
          </Form>

          <S.Section>
            <S.Heading>Endereço</S.Heading>
            <Address addresses={user.address} />
          </S.Section>
        </S.Main>
      </S.Wrapper>
    </Page>
  )
}

export default MyAccount
