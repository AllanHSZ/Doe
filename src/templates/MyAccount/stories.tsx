import { Story, Meta } from '@storybook/react/types-6-0'
import MyAccount from '.'

export default {
  title: 'MyAccount',
  component: MyAccount
} as Meta

export const Default: Story = () => <MyAccount />
