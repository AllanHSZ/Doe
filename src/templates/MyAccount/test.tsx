import { renderWithTheme } from 'utils/tests/helpers'

import MyAccount from '.'

describe('<MyAccount />', () => {
  it('should render the heading', () => {
    renderWithTheme(<MyAccount />)

    expect(true).toBeTruthy()

    // expect(container.firstChild).toMatchSnapshot()
  })
})
