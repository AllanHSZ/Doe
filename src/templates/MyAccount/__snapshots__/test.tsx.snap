// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MyAccount /> should render the heading 1`] = `
.c9 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-family: Lato,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  border: 0;
  cursor: pointer;
  border-radius: 0.6rem;
  padding: 0.8rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #FAFAFA;
  background: #04D361;
  height: 4rem;
  font-size: 1.4rem;
  font-weight: 700;
  padding: 0.8rem 3.2rem;
}

.c9:hover {
  background: #04AB4F;
}

.c13 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-family: Lato,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  border: 0;
  cursor: pointer;
  border-radius: 0.6rem;
  padding: 0.8rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #04D361;
  border: 2px solid currentColor;
  background: transparent;
  height: 4rem;
  font-size: 1.4rem;
  font-weight: 700;
  padding: 0.8rem 3.2rem;
}

.c13:hover {
  color: #04AB4F;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.8rem;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 1.6rem;
  gap: 0.8rem;
  width: 100%;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  font-size: 1.4rem;
  padding: 0.8rem 1.6rem;
  box-shadow: 0px 0px 4px rgba(0,0,0,0.2);
  background: #fff;
  gap: 2.4rem;
}

.c12 svg {
  color: #E33D3D;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: #F2F3F5;
  border-radius: 0.2rem;
  padding: 0 1.6rem;
  border: 0.2rem solid;
  border-color: #F2F3F5;
}

.c6:focus-within {
  box-shadow: 0 0 0.5rem #04D361;
}

.c7 {
  color: #394338;
  font-family: Lato,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.4rem;
  padding: 0.8rem 0;
  padding-left: 1.6rem;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c7:-webkit-autofill {
  -webkit-box-shadow: 0 0 0 2.4rem #F2F3F5 inset;
  -webkit-filter: none;
  filter: none;
}

.c5 {
  font-size: 1.4rem;
  color: #394338;
  margin-bottom: 0.4rem;
  cursor: pointer;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-top: 1.6rem;
  gap: 0.8rem;
  max-width: 32rem;
}

.c0 input {
  width: 100%;
}

.c1 {
  margin-top: 2.4rem;
}

.c8 {
  margin-bottom: 0.8rem;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 2.4rem 0;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  gap: 2.4rem;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.8rem;
  width: 100%;
}

<div
  data-testid="Mock Page"
>
  <div
    class="c0"
  >
    <h3
      class="c1"
    >
      Minha conta
    </h3>
    <main
      class="c2"
    >
      <section
        class="c3"
      >
        <div
          class="c4"
        >
          <label
            class="c5"
          >
            Nome
          </label>
          <div
            class="c6"
          >
            <input
              class="c7"
              type="text"
              value=""
            />
          </div>
        </div>
      </section>
      <section
        class="c3"
      >
        <h5
          class="c8"
        >
          Alterar senha
        </h5>
        <div
          class="c4"
        >
          <label
            class="c5"
          >
            Senha
          </label>
          <div
            class="c6"
          >
            <input
              class="c7"
              type="password"
              value=""
            />
          </div>
        </div>
        <div
          class="c4"
        >
          <label
            class="c5"
          >
            Confirmar senha
          </label>
          <div
            class="c6"
          >
            <input
              class="c7"
              type="password"
              value=""
            />
          </div>
        </div>
      </section>
      <button
        class="c9"
      >
        <span>
          Salvar
        </span>
      </button>
      <section
        class="c3"
      >
        <h5
          class="c8"
        >
          Endere√ßo
        </h5>
        <div
          class="c10"
        >
          <div
            class="c11"
          >
            <div
              class="c12"
            >
              <p>
                Av. XXX, 000 - Bairro Xx
              </p>
              <svg
                fill="none"
                height="18"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="18"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polyline
                  points="3 6 5 6 21 6"
                />
                <path
                  d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
                />
              </svg>
            </div>
            <div
              class="c12"
            >
              <p>
                Av. XXX, 000 - Bairro Xx
              </p>
              <svg
                fill="none"
                height="18"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="18"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polyline
                  points="3 6 5 6 21 6"
                />
                <path
                  d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
                />
              </svg>
            </div>
          </div>
          <button
            class="c13"
          >
            <span>
              Adicionar
            </span>
          </button>
        </div>
      </section>
    </main>
  </div>
</div>
`
